_posts/2022-04-24-ui-architecture.md

```diff
@@ -0,0 +1,264 @@
+
+```rust
+VStack<u32, (), (String, Button<u32, (), {anonymous function of type FnMut(u32) -> ()}>)>
+```
+
+For such a simple example, this is not too bad (other than the callback), but would get annoying quickly.
+
+Another observation is that three nodes of this tree implement the View trait: VStack and its two children String and Button. Yes, that's an ordinary Rust string, and it implements the View trait. So do colors and shapes (following SwiftUI). [TODO: colors and shapes are not implemented yet, but might be by the time the blog is published]
+
+The View trait has two associated types, one for view state, the other for the associated widget tree. In this case, neither is particularly interesting, the view state just has the id's of the two child nodes of the vstack (and more about id's later), and the widget is just `widget::VStack`. All these types are fully inferred by the Rust type system, the programmer rarely has to interact with any of them directly.
+
+On first run of the app logic, the UI runtime *builds* the associated widget tree from the returned view tree. That makes the following hierarchy:
+
+```
+       +--------+
+       | VStack |
+       +--------+
+        /      \
+       /        \
+   +------+   +--------+
+   | Text |   | Button |
+   +------+   +--------+
+```
+
+### Identity and id paths
+
+A specific detail when building the widget tree is assigning a *stable identity* to each widget. These concepts are explained pretty well in the [Demystify SwiftUI] talk. As in SwiftUI, stable identity can be based on *structure* (views in the same place in the view tree get to keep their identity across runs), or an *explicit key.* To illustrate the latter, assume a list container, and that two of the elements in the list are swapped. That might play an animation of the visual representations of those two elements changing places.
+
+Thus, our running example might look like this, annotated with ids:
+
+```
+       1
+       +--------+
+       | VStack |
+       +--------+
+        /      \
+   2   /      3 \
+   +------+   +--------+
+   | Text |   | Button |
+   +------+   +--------+
+```
+
+The idea of assigning a stable identity to a widget is quite standard in declarative UI (it's also present in basically all non-toy immediate mode GUI implementations), but Xilem adds a distinctive twist, the use of *id path* rather than a single id. The id path of a widget is the sequence of all ids on the path from the root to that widget in the widget tree. Thus, the id path of the button in the above is `[1, 3]`, while the label is `[1, 2]` and the stack is just `[1]`. The full id path is redundant if we had global information about the structure of the tree (for example, by following parent links), but the point is that given id paths, we don't *need* to track this kind of information.
```

Comment:
I added something, but I think you've asked a deep question here, and I'm not positive I can give a good argument at the level of writing I'm doing.

I did indeed in an earlier prototype have tree structure tracking that was done on construction/mutation of the widget tree, so the is visible during reconciliation were just plain integer tokens (not paths). One of the subtle ways that breaks down is ids allocated to futures - they might not correspond to an actual node in the widget tree, or perhaps dummy nodes would be added. Same for env_get, which I've written about on Zulip but not in this blog.

Note also that Elm has an analogous problem; Html has `map` but there is no comparable component-forming operation for subscriptions. As far as I know, those have to be addressed to the top-level update method of the app logic.

I'll think a little more about whether I can explain things better. Most often, when I find things hard to explain, it's because of a gap in my own understanding.